# Grouping and Aggregation

## Contains:
### Mean by department_id
### Mean price of item by loyalty flag
### Creation of spending flag column
### Creation of frequency flag column
### Cleaning outliers from item prices

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#set up file path
path = r'C:\Users\krist\08.2023_InstacartBasketAnalysis'

In [3]:
#import dataframe from reading and check import
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02.Data', 'Prepared_Data', 'orders_products_merged_4.8_reading.pkl'))
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                   11.0         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

   ... aisle_id  department_id  prices  _exists    price_range_loc  \
0  ...       77              7     9.0     both  Mid-range product   
1  ...       77              7     9.0     both  Mid-range product   
2  ...       77              7     9.0     both  Mid-range product   
3  ...       77              7     9.0     both  Mid-range product   
4  ...       77              7     9.0     both  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day max_order  \
0  Regularly busy   Regularly busy        average orders        10   
1  Regularly busy  Least busy days        average orders        10   
2  Regularly busy  Least busy days           most orders        10   
3      Least busy  Least busy days        average orders        10   
4      Least busy  Least busy days           most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 21 columns]

In [4]:
#check for shape
df_ords_prods_merge.shape

(32404859, 21)

### shape matches df exported from 4.8 reading: 324,004,859 rows 21 columns

In [5]:
#find aggregated mean of order_number grouped by department_id
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Checked script from 4.4 for departments
### the lowest number of orders per department is 15.2 in alcohol (dept 5)
### these departments all have 15 orders frozen, alcohol, pets, dry goods pasta, meat seafood, and household (1,5,8,9,12,17)
### the highest number of orders per department is 22.9 in missing (dept 21), the next highest is in bulk (dept 10) 
### in the subset we didn't have all departments and the lowest number of orders was 11 in household (dept 17)
### in the subset we didn't have all departments and the highest number of orders was 19.4 in dairy eggs (dept 16)
### if we had only used the subset we would have had the wrong highs and low and the wrong departments 

In [6]:
#loyalty flag already exists because I imported the dataset from the reading exercise
df_ords_prods_merge['loyalty_flag'].head(25)

0         New customer
1         New customer
2         New customer
3         New customer
4         New customer
5         New customer
6         New customer
7         New customer
8         New customer
9         New customer
10    Regular customer
11    Regular customer
12    Regular customer
13    Regular customer
14    Regular customer
15        New customer
16        New customer
17        New customer
18    Regular customer
19    Regular customer
20    Regular customer
21    Regular customer
22    Regular customer
23    Regular customer
24    Regular customer
Name: loyalty_flag, dtype: object

In [7]:
#check to see if type of customer impacts average price of purchased products
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

### on average loyal customers by products that are less expensive than new or regular customeres
### loyal customer product price mean = 10.39
### regular customer product price mean = 12.50
### new customer product price mean = 13.29

## Create new flag by first grouping by user and finding avg product price for each user then creating flag column

In [8]:
#create a new column with mean of product price for each user
df_ords_prods_merge['users_avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [9]:
#check for new column
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                   11.0         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

   ... department_id  prices  _exists    price_range_loc     busiest_day  \
0  ...             7     9.0     both  Mid-range product  Regularly busy   
1  ...             7     9.0     both  Mid-range product  Regularly busy   
2  ...             7     9.0     both  Mid-range product  Regularly busy   
3  ...             7     9.0     both  Mid-range product      Least busy   
4  ...             7     9.0     both  Mid-range product      Least busy   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        average orders        10  New customer   
1  Least busy days        average orders        10  New customer   
2  Least busy days           most orders        10  New customer   
3  Least busy days        average orders        10  New customer   
4  Least busy days           most orders        10  New customer   

   users_avg_product_price  
0                 6.367797  
1                 6.367797  
2                 6.367797  
3                 6.367797  
4                 6.367797  

[5 rows x 22 columns]

In [10]:
#create low spending flag with loc()
df_ords_prods_merge.loc[df_ords_prods_merge['users_avg_product_price'] < 10, 'Spending_flag'] = 'Low spender'

In [11]:
#create high spending flag with loc()
df_ords_prods_merge.loc[df_ords_prods_merge['users_avg_product_price'] >= 10, 'Spending_flag'] = 'High spender'

In [12]:
#check frequencies of new flag
df_ords_prods_merge['Spending_flag'].value_counts()

Low spender     31770614
High spender      634245
Name: Spending_flag, dtype: int64

### high spender count 634,245
### low spender count 31,770,614

In [13]:
#check columns used to create flag
df_ords_prods_merge[['user_id', 'users_avg_product_price', 'Spending_flag']].head(100)

user_id  users_avg_product_price Spending_flag
0         1                 6.367797   Low spender
1         1                 6.367797   Low spender
2         1                 6.367797   Low spender
3         1                 6.367797   Low spender
4         1                 6.367797   Low spender
..      ...                      ...           ...
95      360                10.006250  High spender
96      377                 8.496552   Low spender
97      387                 7.396610   Low spender
98      420                 7.387805   Low spender
99      420                 7.387805   Low spender

[100 rows x 3 columns]

In [14]:
#show all rows to view output
pd.options.display.max_rows = None
df_ords_prods_merge[['user_id', 'users_avg_product_price', 'Spending_flag']].head(100)

user_id  users_avg_product_price Spending_flag
0         1                 6.367797   Low spender
1         1                 6.367797   Low spender
2         1                 6.367797   Low spender
3         1                 6.367797   Low spender
4         1                 6.367797   Low spender
5         1                 6.367797   Low spender
6         1                 6.367797   Low spender
7         1                 6.367797   Low spender
8         1                 6.367797   Low spender
9         1                 6.367797   Low spender
10       15                 3.980556   Low spender
11       15                 3.980556   Low spender
12       15                 3.980556   Low spender
13       15                 3.980556   Low spender
14       15                 3.980556   Low spender
15       19                 8.372059   Low spender
16       19                 8.372059   Low spender
17       19                 8.372059   Low spender
18       21                 7.012683   Low spender
19       31                 8.137793   Low spender
20       31                 8.137793   Low spender
21       43                 7.444898   Low spender
22       43                 7.444898   Low spender
23       52                 6.356287   Low spender
24       52                 6.356287   Low spender
25       52                 6.356287   Low spender
26       52                 6.356287   Low spender
27       52                 6.356287   Low spender
28       52                 6.356287   Low spender
29       52                 6.356287   Low spender
30       52                 6.356287   Low spender
31       52                 6.356287   Low spender
32       52                 6.356287   Low spender
33       52                 6.356287   Low spender
34       52                 6.356287   Low spender
35       52                 6.356287   Low spender
36       52                 6.356287   Low spender
37       67                 7.656790   Low spender
38       67                 7.656790   Low spender
39       67                 7.656790   Low spender
40       67                 7.656790   Low spender
41       67                 7.656790   Low spender
42       67                 7.656790   Low spender
43       67                 7.656790   Low spender
44       67                 7.656790   Low spender
45       67                 7.656790   Low spender
46       67                 7.656790   Low spender
47       67                 7.656790   Low spender
48       67                 7.656790   Low spender
49       67                 7.656790   Low spender
50       67                 7.656790   Low spender
51       67                 7.656790   Low spender
52       67                 7.656790   Low spender
53       67                 7.656790   Low spender
54       67                 7.656790   Low spender
55       67                 7.656790   Low spender
56       81                 8.165385   Low spender
57       81                 8.165385   Low spender
58       82                 6.441221   Low spender
59       82                 6.441221   Low spender
60       82                 6.441221   Low spender
61       82                 6.441221   Low spender
62       82                 6.441221   Low spender
63       82                 6.441221   Low spender
64       82                 6.441221   Low spender
65       82                 6.441221   Low spender
66       82                 6.441221   Low spender
67       98                 8.028000   Low spender
68       98                 8.028000   Low spender
69       98                 8.028000   Low spender
70       98                 8.028000   Low spender
71       98                 8.028000   Low spender
72       98                 8.028000   Low spender
73       98                 8.028000   Low spender
74      109                 5.850725   Low spender
75      120                 9.385714   Low spender
76      120                 9.385714   Low spender
77      185               

## Create a frequency flag by grouping by user_id and finding median of days_last_prior_order

In [15]:
#create new column for median 
df_ords_prods_merge['median_days_since_last_order'] = 
    df_ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [16]:
#check output
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                   11.0         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

   ... _exists    price_range_loc     busiest_day     busiest_days  \
0  ...    both  Mid-range product  Regularly busy   Regularly busy   
1  ...    both  Mid-range product  Regularly busy  Least busy days   
2  ...    both  Mid-range product  Regularly busy  Least busy days   
3  ...    both  Mid-range product      Least busy  Least busy days   
4  ...    both  Mid-range product      Least busy  Least busy days   

  busiest_period_of_day max_order  loyalty_flag users_avg_product_price  \
0        average orders        10  New customer                6.367797   
1        average orders        10  New customer                6.367797   
2           most orders        10  New customer                6.367797   
3        average orders        10  New customer                6.367797   
4           most orders        10  New customer                6.367797   

  Spending_flag  median_days_since_last_order  
0   Low spender                          20.0  
1   Low spender                          20.0  
2   Low spender                          20.0  
3   Low spender                          20.0  
4   Low spender                          20.0  

[5 rows x 24 columns]

In [17]:
#create frequency flag
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_since_last_order'] >20, 'frequency_flag'] = 'Non-frequent customer'

In [18]:
#create frequency flag
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_since_last_order'] <=20) &
                        (df_ords_prods_merge['median_days_since_last_order'] >10), 'frequency_flag'] = 'Regular customer'

In [19]:
#create frequency flag
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_since_last_order'] <=10, 'frequency_flag'] = 'Frequent customer'

In [20]:
#check columns used to create flag
df_ords_prods_merge[['median_days_since_last_order', 'frequency_flag']].head(25)

median_days_since_last_order     frequency_flag
0                           20.0   Regular customer
1                           20.0   Regular customer
2                           20.0   Regular customer
3                           20.0   Regular customer
4                           20.0   Regular customer
5                           20.0   Regular customer
6                           20.0   Regular customer
7                           20.0   Regular customer
8                           20.0   Regular customer
9                           20.0   Regular customer
10                          10.5   Regular customer
11                          10.5   Regular customer
12                          10.5   Regular customer
13                          10.5   Regular customer
14                          10.5   Regular customer
15                           7.0  Frequent customer
16                           7.0  Frequent customer
17                           7.0  Frequent customer
18                           8.0  Frequent customer
19                           6.0  Frequent customer
20                           6.0  Frequent customer
21                          10.0  Frequent customer
22                          10.0  Frequent customer
23                           8.0  Frequent customer
24                           8.0  Frequent customer

In [21]:
#check frequencies of frequency flag
df_ords_prods_merge['frequency_flag'].value_counts()

Frequent customer        20550144
Regular customer          9154455
Non-frequent customer     2700260
Name: frequency_flag, dtype: int64

### frequent customers = 20,550,144
### regular customers = 9,154,455
### non-frequent customers = 2,700,260
### most customers are considered frequent customers, place orders 10 or fewer days apart

## Export file

In [27]:
#export file to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02.Data', 'Prepared_Data', 'orders_products_merged_4.8_task.pkl'))